binr/r2r/run.c

```diff
@@ -39,48 +39,59 @@ struct r2r_subprocess_t {
 
 static RPVector subprocs;
 static RThreadLock *subprocs_mutex;
+static int sigchld_pipe[2];
+static RThread *sigchld_thread;
 
-static void subprocs_lock(sigset_t *old_sigset) {
-	sigset_t block_sigset;
-	sigemptyset (&block_sigset);
-	sigaddset (&block_sigset, SIGWINCH);
-	r_signal_sigmask (SIG_BLOCK, &block_sigset, old_sigset);
-	r_th_lock_enter (subprocs_mutex);
-}
-
-static void subprocs_unlock(sigset_t *old_sigset) {
-	r_th_lock_leave (subprocs_mutex);
-	r_signal_sigmask (SIG_SETMASK, old_sigset, NULL);
+static void handle_sigchld() {
+	ut8 b = 1;
+	write (sigchld_pipe[1], &b, 1);
 }
 
-static void handle_sigchld() {
+static RThreadFunctionRet sigchld_th(RThread *th) {
 	while (true) {
-		int wstat;
-		pid_t pid = waitpid (-1, &wstat, WNOHANG);
-		if (pid <= 0)
-			return;
-
-		void **it;
-		R2RSubprocess *proc = NULL;
-		r_pvector_foreach (&subprocs, it) {
-			R2RSubprocess *p = *it;
-			if (p->pid == pid) {
-				proc = p;
-				break;
+		ut8 b;
+		ssize_t rd = read (sigchld_pipe[0], &b, 1);
```

Comment:
this is not portable

Reply:
This part of the code is for unix only, so should be fine.